qt-gui-cpp build broken on windows

### Solution to issue cannot be found in the documentation.

- [X] I checked the documentation.

### Issue

Let's continue the conversation from #295 here.

Latest error (from https://github.com/RoboStack/ros-noetic/actions/runs/3991408782/jobs/6846172048):
```
link /NOLOGO /DYNAMICBASE /NXCOMPAT /OPT:REF /INCREMENTAL:NO /DLL /SUBSYSTEM:WINDOWS /OUT:release\libqt_gui_cpp_sip.dll @C:\Users\RUNNER~1\AppData\Local\Temp\nm385D.tmp
[6797](https://github.com/RoboStack/ros-noetic/actions/runs/3991408782/jobs/6846172048#step:6:6798)
   Creating library release\libqt_gui_cpp_sip.lib and object release\libqt_gui_cpp_sip.exp
[6798](https://github.com/RoboStack/ros-noetic/actions/runs/3991408782/jobs/6846172048#step:6:6799)
siplibqt_gui_cpp_sipqt_gui_cppRosPluginlibPluginProvider_ForPlugins.obj : error LNK2019: unresolved external symbol "bool ros::console::g_initialized" (?g_initialized@console@ros@@3_NA) referenced in function "public: __cdecl pluginlib::ClassLoader<class qt_gui_cpp::Plugin>::ClassLoader<class qt_gui_cpp::Plugin>(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,class std::vector<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,class std::allocator<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > > >)" (??0?$ClassLoader@VPlugin@qt_gui_cpp@@@pluginlib@@QEAA@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@00V?$vector@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@V?$allocator@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@2@@3@@Z)
[6799](https://github.com/RoboStack/ros-noetic/actions/runs/3991408782/jobs/6846172048#step:6:6800)
siplibqt_gui_cpp_sipqt_gui_cppRosPluginlibPluginProvider_ForPluginProviders.obj : error LNK2001: unresolved external symbol "bool ros::console::g_initialized" (?g_initialized@console@ros@@3_NA)
[6800](https://github.com/RoboStack/ros-noetic/actions/runs/3991408782/jobs/6846172048#step:6:6801)
siplibqt_gui_cpp_sipqt_gui_cppPluginBridge.obj : error LNK2001: unresolved external symbol "public: static struct QMetaObject const qt_gui_cpp::PluginBridge::staticMetaObject" (?staticMetaObject@PluginBridge@qt_gui_cpp@@2UQMetaObject@@B)
[6801](https://github.com/RoboStack/ros-noetic/actions/runs/3991408782/jobs/6846172048#step:6:6802)
siplibqt_gui_cpp_sipqt_gui_cppPlugin.obj : error LNK2001: unresolved external symbol "public: static struct QMetaObject const qt_gui_cpp::Plugin::staticMetaObject" (?staticMetaObject@Plugin@qt_gui_cpp@@2UQMetaObject@@B)
[6802](https://github.com/RoboStack/ros-noetic/actions/runs/3991408782/jobs/6846172048#step:6:6803)
release\libqt_gui_cpp_sip.dll : fatal error LNK1120: 3 unresolved externals
[6803](https://github.com/RoboStack/ros-noetic/actions/runs/3991408782/jobs/6846172048#step:6:6804)
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30133\bin\HostX64\x64\link.EXE"' : return code '0x460'
[6804](https://github.com/RoboStack/ros-noetic/actions/runs/3991408782/jobs/6846172048#step:6:6805)
Stop.
[6805](https://github.com/RoboStack/ros-noetic/actions/runs/3991408782/jobs/6846172048#step:6:6806)
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30133\bin\HostX64\x64\nmake.exe"' : return code '0x2'
[6806](https://github.com/RoboStack/ros-noetic/actions/runs/3991408782/jobs/6846172048#step:6:6807)
Stop.
[6807](https://github.com/RoboStack/ros-noetic/actions/runs/3991408782/jobs/6846172048#step:6:6808)
NMAKE : fatal error U1077: 'cd' : return code '0x2'
[6808](https://github.com/RoboStack/ros-noetic/actions/runs/3991408782/jobs/6846172048#step:6:6809)
Stop.
[6809](https://github.com/RoboStack/ros-noetic/actions/runs/3991408782/jobs/6846172048#step:6:6810)
Error: could not load cache
```

/cc @traversaro do you have any clue?

### Installed packages

```shell
n/a
```


### Environment info

```shell
n/a
```

--- a/vinca/templates/bld_catkin.bat.in
+++ b/vinca/templates/bld_catkin.bat.in
@@ -6,7 +6,7 @@ set "PYTHONPATH=%LIBRARY_PREFIX%\lib\site-packages;%SP_DIR%"
 :: MSVC is preferred.
 set CC=cl.exe
 set CXX=cl.exe
-https://learn.microsoft.com/en-us/cpp/build/reference/cl-environment-variables?view=msvc-170
+
 :: ROS_BUILD_SHARED_LIBS is always defined in CMake by catkin
 :: if ROS (1) is build as shared library . However, some packages are not
 :: passing compilation flags from CMake to other build systems (such as qmake),

--- a/vinca/templates/bld_catkin.bat.in
+++ b/vinca/templates/bld_catkin.bat.in
@@ -6,6 +6,13 @@ set "PYTHONPATH=%LIBRARY_PREFIX%\lib\site-packages;%SP_DIR%"
 :: MSVC is preferred.
 set CC=cl.exe
 set CXX=cl.exe
+https://learn.microsoft.com/en-us/cpp/build/reference/cl-environment-variables?view=msvc-170
+:: ROS_BUILD_SHARED_LIBS is always defined in CMake by catkin
+:: if ROS (1) is build as shared library . However, some packages are not
+:: passing compilation flags from CMake to other build systems (such as qmake),
+:: so we enable it explicitly via the CL environment variable, see
+:: https://learn.microsoft.com/en-us/cpp/build/reference/cl-environment-variables?view=msvc-170
+set CL=/DROS_BUILD_SHARED_LIBS=1
 
 set "CATKIN_BUILD_BINARY_PACKAGE_ARGS=-DCATKIN_BUILD_BINARY_PACKAGE=1"
 if "%PKG_NAME%" == "ros-@(ros_distro)-catkin" (

--- a/vinca/templates/bld_catkin_merge.bat.in
+++ b/vinca/templates/bld_catkin_merge.bat.in
@@ -6,6 +6,13 @@ setlocal
 set CC=cl.exe
 set CXX=cl.exe
 
+:: ROS_BUILD_SHARED_LIBS is always defined in CMake by catkin
+:: if ROS (1) is build as shared library . However, some packages are not
+:: passing compilation flags from CMake to other build systems (such as qmake),
+:: so we enable it explicitly via the CL environment variable, see
+:: https://learn.microsoft.com/en-us/cpp/build/reference/cl-environment-variables?view=msvc-170
+set CL=/DROS_BUILD_SHARED_LIBS=1
+
 set CATKIN_MAKE_ISOLATED=src\ros-@(ros_distro)-catkin\bin\catkin_make_isolated
 set CMAKE_PREFIX_PATH=%CMAKE_PREFIX_PATH:\=/%
 

